// @aws-sdk/retry-config-provider/dist/cjs/defaultProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.retryModeProvider = exports.maxAttemptsProvider = exports.CONFIG_RETRY_MODE = exports.ENV_RETRY_MODE = exports.CONFIG_MAX_ATTEMPTS = exports.ENV_MAX_ATTEMPTS = void 0;
var middleware_retry_1 = require("@aws-sdk/middleware-retry");
var property_provider_1 = require("@aws-sdk/property-provider");
var fromEnv_1 = require("./fromEnv");
var fromSharedConfigFiles_1 = require("./fromSharedConfigFiles");
exports.ENV_MAX_ATTEMPTS = "AWS_MAX_ATTEMPTS";
exports.CONFIG_MAX_ATTEMPTS = "max_attempts";
exports.ENV_RETRY_MODE = "AWS_RETRY_MODE";
exports.CONFIG_RETRY_MODE = "retry_mode";
var defaultProvider = function (configuration, _a) {
    if (configuration === void 0) { configuration = {}; }
    var envVarName = _a.envVarName, configKey = _a.configKey, defaultValue = _a.defaultValue;
    return property_provider_1.memoize(property_provider_1.chain(fromEnv_1.fromEnv(envVarName), fromSharedConfigFiles_1.fromSharedConfigFiles(configuration, configKey), property_provider_1.fromStatic(defaultValue)));
};
exports.maxAttemptsProvider = function (configuration) {
    if (configuration === void 0) { configuration = {}; }
    return defaultProvider(configuration, {
        envVarName: exports.ENV_MAX_ATTEMPTS,
        configKey: exports.CONFIG_MAX_ATTEMPTS,
        defaultValue: middleware_retry_1.DEFAULT_MAX_ATTEMPTS,
    });
};
exports.retryModeProvider = function (configuration) {
    if (configuration === void 0) { configuration = {}; }
    return defaultProvider(configuration, {
        envVarName: exports.ENV_RETRY_MODE,
        configKey: exports.CONFIG_RETRY_MODE,
        defaultValue: middleware_retry_1.DEFAULT_RETRY_MODE,
    });
};
